[sdk-54][iOS][android] Support precompiled React Native binaries in expo-build-props

chrfalch opened the pull request:
# Why

The property buildReactNativeFromSource in app.json correctly sets the env variable for using React Native precompiled dependencies available from 0.80.

# How

- Renamed setting from iOS.buildFromSource -> iOS.buildReactNativeFromSource
- Deprecated iOS.buildFromSource
- Setting the environment variable for using precompiled React Native binaries as well.
- Docs updated

# Test Plan

Create new SDK-54 project and observe:
- app.json has `buildReactNativeFromSource` = false
- Podfile has `RCT_USE_PREBUILT_RNCORE` set if buildFromSource is false

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Follows documentation rules

chrfalch committed:
[sdk-54] Support precompiled React Native binaries in app.json

The property buildFromSource in app.json correctly sets the env variable for using React Native precompiled dependencies available from 0.80.

This commit adds support for setting the environment variable for using precompiled React Native binaries as well.

In addition new projects will get app.json files with this property set to false so that for new projects we'll enable this by default.

chrfalch committed:
Update packages/expo-build-properties/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

gabrieldonadel reviewed (commented):
Given that we already support `ios.buildFromSource` through expo-build-properties via https://github.com/expo/expo/pull/37678, I think we just need to change the Pofile of expo-tempalte-bare-mininum to enable it by default

chrfalch committed:
codereview: reverted changes

chrfalch committed:
codrereview: swapped behavior + cleanup name

- Changed from ios.buildFromSource -> ios.buildReactNativeFromSource
- Deprecated ios.buildFromSource
- True by default

chrfalch committed:
Reverted commit

chrfalch committed:
[android] Renamed buildFromSource -> buildReactNativeFromSource

This commit renames the above on Android and deprecates the old name

chrfalch committed:
Changelog entry update

gabrieldonadel reviewed (approved):
Looks good to me, just one thing

chrfalch committed:
Update packages/expo-build-properties/src/android.ts

Co-authored-by: Gabriel Donadel Dall'Agnol <gabriel@expo.io>

chrfalch committed:
Rebuild expo-build-properties

chrfalch committed:
Updated unit tests

chrfalch committed:
Added test to avoid enabling precompiled with old arc

Old arch shouldn't allow using precompiled.

- Updated docs
- Updated podfile
- Rebuilt docs + types
